pandas/tests/io/json/test_json_table_schema.py

```diff
@@ -571,3 +561,11 @@ def test_multiindex(self, index_names):
         out = df.to_json(orient="table")
         result = pd.read_json(out, orient="table")
         tm.assert_frame_equal(df, result)
+    
+    def test_empty_frame_roundtrip(self):
+        # GH 21287
+        df = pd.DataFrame([], columns=['a', 'b', 'c'])
+        expected = df.copy()
+        out = df.to_json(None, orient='table')
+        result = pd.read_json(out, orient='table')
+        tm.assert_frame_equal(expected, result)
```

Comment:
This raises an assertion error:
```
E       AssertionError: DataFrame.index are different
E
E       DataFrame.index classes are not equivalent
E       [left]:  Index([], dtype='object')
E       [right]: Float64Index([], dtype='float64')
```

That's something I need to dig deeper, but any pointers on what might be the reason for this would be great.

Reply:
Thanks for doing this PR! Beat me to it :)
A bit weak but what do we think of just
`pd.testing.assert_frame_equal(expected, actual, check_dtype=False)` ?

Otherwise I would guess we have to go down the road of including the dtypes in the JSON representation?

Reply:
Good point! And actually it only works if I assert it like this:
`pd.testing.assert_frame_equal(expected, result, check_dtype=False, check_index_type=False)`
So both `check_dtype=False`and `check_index_type=False` have to be set to `False` in order to get the assertion right.

Thoughts on this?

Reply:
Ignoring the dtype difference is not the solution. The point of this format is to persist that metadata.

What I would do is check that the proper type information for the index is being written out (you can use a io.StringIO instance instead of writing to `None`). If that appears correct then there would be an issue with the reader that is ignoring or casting the type of the index after the fact

Reply:
Yeah, you're right, ignoring the `dtype` and the `index_type` will just hide the problem.

Did some initial testings and it seems that on the reading side empty `data` with `data.dtype == 'object'` gets coerced to Float64 without any clear reason.

I'll push a commit with fix proposal for comments.